File: legacy-python/geocoder.py
#!/usr/bin/python
#
# Copyright 2011 Foursquare Labs Inc. All Rights Reserved.

import csv
import functools
import pymongo
import sys
import json

from geotools import *
from optparse import OptionParser

from pymongo import Connection
connection = Connection()
db = connection.geonames_global2
geonames = db.features

# todo
# online deduping
# highlighting?
# if I were me, I would add hacks for us+4, canadian+3 and GB+3 zipcodes

class Geocoder:
  def __init__(self, tryHard=True):
    self.tryHard = tryHard
    
  def find_name(self, tokens):
    name = ' '.join(tokens)
    #print 'monging for: %s' % name
    features = geonames.find({'names': name})
    return [f for f in features]

  def generate_parses(self, tokens, cache):
    if len(tokens) == 0:
      return [[]]

    # better cache key pleez
    cache_key = len(tokens)
    if cache_key not in cache:
      parses = []
      for i in xrange(1, len(tokens) + 1):
        #print "trying: %d to %d: %s" % (0, i, ' '.join(tokens[0:i]))
        features = self.find_name(tokens[0:i])
        #print "have %d matches" % len(features)
        sub_parses = self.generate_parses(tokens[i:], cache)
        for f in features:
          #print "feature: %s" % f
          for p in sub_parses:
            #print "sub_parse: %s" % p
            parse = [f] + p
            parse = sorted(parse, key=self.feature_comparator)
            if self.is_valid(parse):
              #print "Valud"
              parses.append(parse)

      #print "setting %s" % cache_key
      cache[cache_key] = parses

    return cache[cache_key]


  # this should really use the geonames class hierarchy, not parent len
  def feature_comparator(self, feature):
    if feature['feature_class'] == 'ZIP':
      return -10000
    else:
      return -1*len(feature.get('parents', []))

  def is_valid(self, parse):
    # parse is a list of features 
    if self.is_valid_helper(parse):
      return True
    else:
      #was invalid, so try zip fuzzing logic
      most_specific = parse[0]
      if most_specific['feature_class'] == 'ZIP':
        is_valid_without_zip = self.is_valid_helper(parse[1:])
        if is_valid_without_zip:
          # is our zip < 200km from the next smallest feature
          dist = distance_km(most_specific['lat'], most_specific['lng'], parse[1]['lat'], parse[1]['lng'])
          #print "%d km from %s to %s %s" % (dist, most_specific['names'][0], parse[1]['names'][0], parse[1]['featureid'])
          #print "%s,%s to %s,%s" % (most_specific['lat'], most_specific['lng'], parse[1]['lat'], parse[1]['lng'])
          if dist < 200:
            return True

      return False
      

  def is_valid_helper(self, parse): 
    most_specific = parse[0]
   
    # don't allow "spots" (buildings, etc) without tryHard
    if not self.tryHard and most_specific['feature_class'] == 'S':
      return False
    
    for feature in parse[1:]:
      if feature['_id'] == most_specific['_id']:
        continue
      if feature['cc'] != most_specific['cc']:
        return False
      if 'featureid' not in feature:
        return False
      if 'parents' not in most_specific:
        return False
      if feature['featureid'] not in most_specific['parents']:
        return False

    return True

  def filter_valid(self, parses):
    return filter(self.is_valid, parses)
 
  def geocode(self, string, ccHint=None, llHint=None, lang=None):
    def rank_parse(parse):
      signal = int(parse[0].get('POP') or 0)
      
      # if we have a repeated feature, downweight this like crazy
      # so st petersburg, st petersburg works
      # but doesn't break new york, ny
      for x in parse[1:]:
        if x['_id'] == parse[0]['_id']:
          signal -= 100000000

      # prefer a more aggressive parse ... bleh
      # this prefers "mt laurel" over "laurel" in "mt"
      signal -= 20000 * len(parse)

      if (ccHint == parse[0]['cc']):
        signal += 100000
      signal += parse[0].get('boost', 0)

      if llHint:
        dist = distance_km(llHint[0], llHint[1], parse[0]['lat'], parse[0]['lng'])
        # tie-break with distance to llHint
        signal -= dist

      # as a tie break, zipcodes in the US > zipcodes elsewhere
      if parse[0]['cc'] == 'US':
        signal += 1

      return -1*signal
 
    tokens = tokenize(normalize(string))
    connector_tokens = []
    query_tokens = []

    # try connectors
    if 'near' in tokens:
      i = tokens.index('near')
      connector_tokens = tokens[i:i+1]
      query_tokens = tokens[0:i]
      tokens = tokens[i+1:]

    cache = {}
    self.generate_parses(tokens, cache)
    best_key = len(tokens)
    if len(cache.keys()) > 0:
      for k in reversed(sorted(cache.keys())):
        if len(cache[k]) > 0:
          best_key = k
          break

    split_point = len(tokens) - best_key
    valid_parses = cache.get(best_key, [])
    sorted_valid_parses = sorted(valid_parses, key=rank_parse)
    connector = ''

    if len(sorted_valid_parses):
      query = ' '.join(tokens[0:split_point])
      geocode = ' '.join(tokens[split_point:])
      if connector_tokens:
        if query != '':
          query = ' '.join(query_tokens + connector_tokens + tokens[0:split_point])
        else:
          query = ' '.join(query_tokens)
          connector = ' '.join(connector_tokens)
    else:
      geocode = ''
      query = ' '.join(tokens)

    return (sorted_valid_parses[:5], {
      'query': query,
      'geocode': geocode,
      'connector': connector
    })

import tornado.ioloop
import tornado.web

class MainHandler(tornado.web.RequestHandler):
  def makeInterpretation(self, meta, geo, parentDict, ccHint=None, verbose=False, lang=None):
    resp = {
      'what': meta['query'],
      'where': meta['geocode'],
      'connector': meta['connector'],
    }
 
    for f in geo:
      f['_id'] = unicode(f['_id'])

    if len(geo):
      resp['cc'] = geo[0]['cc']
      resp['lat'] = geo[0]['lat']
      resp['lng'] = geo[0]['lng']
      if 'bb' in geo[0]:
        resp['geometry'] = {
          'bounds': geo[0]['bb']
        }
      if 'geonameid' in geo[0]:
        resp['geonameid'] = geo[0]['geonameid']

    parents = []
    if 'parents' in geo[0]:
      for p in geo[0]['parents']:
        if p in parentDict:
          parent = parentDict[p]
          parent['_id'] = unicode(parent['_id'])
          parents.append(parent)

    parents = sorted(parents, key=featureComparator)

    def rankName(preferAbbrev, n):
      score = 0
      if n['p']:
        score += 1
      if n['l'] == lang:
        score += 2
      if n['l'] == 'abbr' and preferAbbrev:
        score += 4
      return -1*score

    def getBestName(f, preferAbbrev):
      return sorted(f['displayNames'], key=functools.partial(rankName, preferAbbrev))[0]['s']
    
    bestName = getBestName(geo[0], False)
    names = [bestName]

    for i, p in enumerate(parents):
      if not 'displayNames' in p or not p['displayNames']:
        next
      if p['feature_code'].startswith('PCL') and geo[0]['cc'] == ccHint:
        next
 
      if p['feature_code'].startswith('PCL'):
        names.append(p['cc'])
      else:
        names.append(getBestName(p, True))

    resp['name'] = bestName
    resp['displayString'] = ', '.join(names)
    if verbose:
      resp['feature'] = geo[0]
      resp['parents'] = parents

    return resp

  def get(self):
    def boolArg(name, default):
      argStr = self.get_argument(name, str(default))
      return argStr in ['t', '1', 'true', 'True']
      
    query = self.get_argument("query")
    callback = self.get_argument("callback", None)
   
    tryHard = boolArg("tryHard", True)
    verbose = boolArg("verbose", False)
    
    ccHint = self.get_argument("ccHint", None)
    lang = self.get_argument("lang", "en")
    
    llHintStr = self.get_argument("llHint", None)
    llHint = None
    try:
      llHintParts = llHintStr.split(',')
      llHint = (float(llHintParts[0]), float(llHintParts[1]))
    except:
      pass

    geocoder = Geocoder(tryHard=tryHard)
    (g, meta) = geocoder.geocode(query, ccHint=ccHint, llHint=llHint, lang=lang)
    #print g
    geocodes = g[:3]

    parentids = []
    for geo in geocodes:
      if 'parents' in geo[0]:
        parentids += geo[0]['parents']
    parents = {}
    for p in geonames.find({'featureid': { '$in': parentids }}):
      parents[p['featureid']] = p
    
    interps = [self.makeInterpretation(meta, geo, parents, ccHint=ccHint, verbose=verbose, lang=lang) for geo in geocodes]
    resp = {'interpretations': interps}
   
    if callback:
      self.write(u'%s(%s)' % (callback, json.dumps(resp)))
    else: 
      self.write(resp)

if __name__ == "__main__":
  usage = ""
  parser = OptionParser(usage=usage)
  parser.add_option("-d", "--debug", dest="debug", action="store_true", default=False, help="auto-reload debug mode")
  (options, args) = parser.parse_args()

  app_settings = { 
    'debug': options.debug
  } 

  application = tornado.web.Application([
      (r"/", MainHandler),
      (r"/static/(.*)", tornado.web.StaticFileHandler, dict(path="static")),
  ], **app_settings)

  application.listen(int(args[0]))
  tornado.ioloop.IOLoop.instance().start()

File: server/src/main/scala/Geocoder.scala
//  Copyright 2012 Foursquare Labs Inc. All Rights Reserved
package com.foursquare.twofish

import com.foursquare.twofish.Implicits._
import com.twitter.util.{Future, FuturePool}
import java.util.concurrent.ConcurrentHashMap
import org.bson.types.ObjectId
import scala.collection.JavaConversions._
import scala.collection.mutable.HashMap

object GeocodeServingFeatureOrdering extends Ordering[GeocodeServingFeature] {
  def compare(a: GeocodeServingFeature, b: GeocodeServingFeature) = {
    YahooWoeTypes.getOrdering(a.feature.woeType) - YahooWoeTypes.getOrdering(b.feature.woeType)
  }
}

class GeocoderImpl(store: GeocodeStorageFutureReadService) extends LogHelper {
  type Parse = Seq[GeocodeServingFeature]
  type ParseSeq = Seq[Parse]
  type ParseCache = ConcurrentHashMap[Int, Future[ParseSeq]]

  /*
    The basic algorithm works like this
    - roughly a depth-first recursive descent parser
    - for every set of tokens from size 1->n, attempt to find a feature with a matching name in our datastore
      (at the first level, for "rego park ny", we'd try rego, rego park, rego park ny)
    - for every feature found matching one of our token sets, recursively try to geocode the remaining tokens
    - return early from a branch of our parse tree if the parse becomes invalid/inconsistent, that is,
      if the feature found does not occur in the parents of the smaller feature found 
      (we would about the parse of "los angeles, new york, united states" when we'd consumed "los angeles"
       its parents were CA & US, and then consumed "new york" because NY is not in the parents of LA)
    - save our work in a map[int -> parses], where the int is the total number of tokens those parses consume.
       we do this because:
       if two separate parses made it to the same point in the input, we don't need to redo the work
         (contrived example: Laurel Mt United States, can both be parsed as "Laurel" in "Mt" (Montana), 
          and "Laurel Mt" (Mountain), both consistent & valid parses. One of those parses would have already
         completely explored "united states" before the second one gets to it)
    - we return the entire cache, which is a little silly. The caller knows to look for the cache for the 
      largest key (the longest parse), and to take all the tokens before that and make the "what" (the
      non-geocoded tokens) in the final interpretation.
   */

  def generateParses(tokens: List[String]): ParseCache = {
    val cache = new ParseCache
    generateParsesHelper(tokens, cache)
    cache
  }

  def generateParsesHelper(tokens: List[String], cache: ParseCache): Future[ParseSeq] = {
    val cacheKey = tokens.size
    if (tokens.size == 0) {
      Future.value(List(Nil))
    } else {
      if (!cache.contains(cacheKey)) {
        val searchStrs: Seq[String] =
          1.to(tokens.size).map(i => tokens.take(i).mkString(" "))

        val featuresFs: Seq[Future[Parse]] =
          for ((searchStr, i) <- searchStrs.zipWithIndex) yield {
            val result: Future[Parse] = store.getByName(searchStr)
            result
          }

         // Compute subparses in parallel (scatter)
         val subParsesFs: Seq[Future[ParseSeq]] =
           1.to(tokens.size).map(i => generateParsesHelper(tokens.drop(i), cache))

         // Collect the results of all the features and subparses (gather)
         val featureListsF: Future[ParseSeq] = Future.collect(featuresFs)
         val subParseSeqsF: Future[Seq[ParseSeq]] = Future.collect(subParsesFs)

         val validParsesF: Future[ParseSeq] =
          for((featureLists, subParseSeqs) <- featureListsF.join(subParseSeqsF)) yield {
            val validParses: ParseSeq = (
              featureLists.zip(subParseSeqs).flatMap({case(features: Parse, subparses: ParseSeq) => {
                (for {
                  f <- features
                  val _ = logger.ifTrace("looking at %s".format(f))
                  p <- subparses
                  val _ = logger.ifTrace("sub_parse: %s".format(p))
                } yield {
                  val parse: Parse = Vector(f) ++ p
                  if (isValidParse(parse)) {
                    logger.ifTrace("VALID -- adding to %d".format(cacheKey))
                    logger.ifTrace("adding parse " + parse)
                    logger.ifTrace("sorted " + parse.sorted(GeocodeServingFeatureOrdering))
                    Some(parse.sorted(GeocodeServingFeatureOrdering))
                  } else {
                    logger.ifTrace("INVALID")
                    None
                  }
                }).flatten
              }})
            )
            validParses.toList
          }

        validParsesF.onSuccess(validParses => {
          logger.ifTrace("setting %d to %s".format(cacheKey, validParses))
        })
        cache(cacheKey) = validParsesF
      }
      cache(cacheKey)
    }
  }

  def isValidParse(parse: Parse): Boolean = {
    if (isValidParseHelper(parse)) {
      true
    } else {
      /*
       * zipcode hack
       * zipcodes don't belong to the political hierarchy, so they don't
       * have the parents you'd expect. Also, people call zipcodes a lot
       * of things other than the official name. As a result, we're going
       * to accept a parse that includes a zipcode if it's within 200km
       * of the next smallest feature
       */
      val sorted_parse = parse.sorted(GeocodeServingFeatureOrdering)
      (for {
        first <- sorted_parse.lift(0)
        second <- sorted_parse.lift(1)
      } yield {
        first.feature.woeType == YahooWoeType.POSTAL_CODE &&
        isValidParseHelper(sorted_parse.drop(1)) &&
        GeoTools.getDistance(
          second.feature.geometry.center.lat,
          second.feature.geometry.center.lng,
          first.feature.geometry.center.lat,
          first.feature.geometry.center.lng) < 200000
      }).getOrElse(false)
    }
  }

  def isValidParseHelper(parse: Parse): Boolean = {
    if (parse.size <= 1) {
      true
    } else {
      val most_specific = parse(0)
      val rest = parse.drop(1)
      rest.forall(f => {
        f.id == most_specific.id ||
        most_specific.scoringFeatures.parents.contains(f.id)
      })
    }
  }

  class FeatureNameComparator(lang: Option[String], preferAbbrev: Boolean) extends Ordering[FeatureName] {
    def compare(a: FeatureName, b: FeatureName) = {
      scoreName(b) - scoreName(a)
    }

    def scoreName(name: FeatureName): Int = {
      var score = 0
      if (name.flags.contains(FeatureNameFlags.PREFERRED)) {
        score += 1
      }
      if (lang.exists(_ == name.lang)) {
        score += 2
      }
      if (name.flags.contains(FeatureNameFlags.ABBREVIATION) && preferAbbrev) {
        score += 4
      }
      score
    }
  }

 def bestName(f: GeocodeFeature, lang: Option[String], preferAbbrev: Boolean): Option[FeatureName] = {
   f.names.sorted(new FeatureNameComparator(lang, preferAbbrev)).headOption
 }
 
  class ParseOrdering(llHint: GeocodePoint, ccHint: String) extends Ordering[Parse] {
    // Higher is better
    def scoreParse(parse: Parse): Int = {
      parse.headOption.map(primaryFeature => {
        val rest = parse.drop(1)
        var signal = primaryFeature.scoringFeatures.population

        // if we have a repeated feature, downweight this like crazy
        // so st petersburg, st petersburg works, but doesn't break new york, ny
        if (rest.contains(primaryFeature)) {
          signal -= 100000000
        }

        // prefer a more aggressive parse ... bleh
        // this prefers "mt laurel" over the town of "laurel" in "mt" (montana)
        signal -= 20000 * parse.length

        // Matching country hint is good
        if (Option(ccHint).exists(_ == primaryFeature.feature.cc)) {
          signal += 100000
        }

        Option(llHint).foreach(ll => {
          signal -= GeoTools.getDistance(ll.lat, ll.lng,
              primaryFeature.feature.geometry.center.lat,
              primaryFeature.feature.geometry.center.lng)
        })

        signal += primaryFeature.scoringFeatures.boost

        // as a terrible tie break, things in the US > elsewhere
        // meant primarily for zipcodes
        if (primaryFeature.feature.cc == "US") {
          signal += 1
        }
        signal
      }).getOrElse(0)
    }

    def compare(a: Parse, b: Parse) = {
      scoreParse(b) - scoreParse(a)
    }
  }

  def hydrateParses(parses: Seq[Parse]): Future[ParseSeq] = {
    val ids: Seq[String] = parses.flatMap(_.map(_.id))
    val oids = ids.map(i => new ObjectId(i))
    store.getByObjectIds(oids).map(geocodeRecordMap =>
      parses.map(parse => {
        parse.flatMap(feature => {
          geocodeRecordMap.get(new ObjectId(feature.id))
        })
      })
    )
  }

  def fixFeature(req: GeocodeRequest, f: GeocodeFeature, parents: Seq[GeocodeServingFeature]) {
    // set name
    val name = bestName(f, Some(req.lang), false).map(_.name).getOrElse("")
    f.setName(name)

    // possibly clear names
    if (!req.full) {
      val names = f.names
      f.setNames(names.filter(n => 
        n.flags.contains(FeatureNameFlags.ABBREVIATION) ||
        n.lang == req.lang ||
        n.lang == "en"
      ))
    }

    // set displayName
    val parentsToUse = parents.filter(p => {
      // if it's a country and it's the same country as the request
      // don't use it
      val sameCountry = p.feature.woeType == YahooWoeType.COUNTRY && p.feature.cc == req.cc
      !sameCountry
    })

    val parentNames = parentsToUse.map(p =>
      bestName(p.feature, Some(req.lang), true).map(_.name).getOrElse(""))
    f.setDisplayName((Vector(name) ++ parentNames).mkString(", "))
  }

  def geocode(req: GeocodeRequest): Future[GeocodeResponse] = {
    val query = req.query

    logger.ifTrace("%s --> %s".format(query, NameNormalizer.normalize(query)))

    val tokens = NameNormalizer.tokenize(NameNormalizer.normalize(query))
    logger.ifTrace("--> %s".format(tokens.mkString("_|_")))

    /// CONNECTOR PARSING GOES HERE
    val cache = generateParses(tokens)
    val futureCache: Iterable[Future[(Int, ParseSeq)]] = cache.map({case (k, v) => {
      Future.value(k).join(v)
    }})

    Future.collect(futureCache.toList).flatMap(cache => {
      val validParseCaches: Iterable[(Int, ParseSeq)] = cache.filter(_._2.nonEmpty)

      if (validParseCaches.size > 0) {
        val longest = validParseCaches.map(_._1).max
        val longestParses = validParseCaches.find(_._1 == longest).get._2

        hydrateParses(longestParses).flatMap(hydratedParses => {
          // TODO: make this configurable
          val sortedParses = hydratedParses.sorted(new ParseOrdering(req.ll, req.cc)).take(3)

          val parentIds = sortedParses.flatMap(_.headOption.toList.flatMap(_.scoringFeatures.parents))
          val parentOids = parentIds.map(i => new ObjectId(i))
          logger.ifTrace("parent ids: " + parentOids)
          store.getByObjectIds(parentOids).map(parentMap => {
            logger.ifTrace(parentMap.toString)

            val what = tokens.take(tokens.size - longest).mkString(" ")
            val where = tokens.drop(tokens.size - longest).mkString(" ")
            logger.ifTrace("%d sorted parses".format(sortedParses.size))

            // need to fix names here
            new GeocodeResponse(sortedParses.map(p => {
            //  p(0).setScoringFeatures(null)
              val feature = p(0).feature
              val sortedParents = p(0).scoringFeatures.parents.flatMap(id => parentMap.get(new ObjectId(id))).sorted(GeocodeServingFeatureOrdering)
              fixFeature(req, feature, sortedParents)
              val interp = new GeocodeInterpretation(what, where, feature)
              interp.setParents(sortedParents.map(parentFeature => {
                val sortedParentParents = parentFeature.scoringFeatures.parents.flatMap(id => parentMap.get(new ObjectId(id))).sorted
                val feature = parentFeature.feature
                fixFeature(req, feature, sortedParentParents)
                feature
              }))
              interp
            }))
          })
        })
      } else {
        Future.value(new GeocodeResponse(Nil))
      }
    })
  }
}
